solidity/contracts/Rewards.sol

```diff
@@ -0,0 +1,611 @@
+
+    // Array representing the percentage of unallocated rewards
+    // available for each reward interval.
+    uint256[] intervalWeights; // percent array
+    // Mapping of interval number to tokens allocated for the interval.
+    uint256[] intervalAllocations;
+
+    // mapping of keeps to booleans.
+    // True if the keep has been used to claim a reward.
+    mapping(bytes32 => bool) claimed;
+
+    // Mapping of interval to number of keeps created in/before the interval
+    mapping(uint256 => uint256) keepsByInterval;
+
+    // Mapping of interval to number of keeps whose rewards have been paid out,
+    // or reallocated because the keep closed unhappily
+    mapping(uint256 => uint256) public intervalKeepsProcessed;
+
+    constructor (
+        uint256 _termLength,
+        address _token,
+        uint256 _minimumKeepsPerInterval,
+        uint256 _firstIntervalStart,
+        uint256[] memory _intervalWeights
+    ) public {
+        token = IERC20(_token);
+        termLength = _termLength;
+        firstIntervalStart = _firstIntervalStart;
+        minimumKeepsPerInterval = _minimumKeepsPerInterval;
+        intervalWeights = _intervalWeights;
+    }
+
+    /// @notice Funds the rewards contract.
+    /// @dev Adds the received amount of tokens
+    /// to `totalRewards` and `unallocatedRewards`.
+    /// May be called at any time, even after allocating some intervals.
+    /// Changes to `unallocatedRewards`
+    /// will take effect on subsequent interval allocations.
+    /// Intended to be used with `approveAndCall`.
+    /// If the reward contract has received tokens outside `approveAndCall`,
+    /// this collects them as well.
+    /// @param _from The original sender of the tokens.
+    /// Must have approved at least `_value` tokens for the rewards contract.
+    /// @param _value The amount of tokens to fund.
+    /// @param _token The token to fund the rewards in.
+    /// Must match the one specified in the rewards contract.
+    function receiveApproval(
+        address _from,
+        uint256 _value,
+        address _token,
+        bytes memory
+    ) public {
+        require(IERC20(_token) == token, "Unsupported token");
+
+        token.safeTransferFrom(_from, address(this), _value);
+
+        uint256 currentBalance = token.balanceOf(address(this));
+        uint256 beforeBalance = totalRewards.sub(paidOutRewards);
+        require(
+            currentBalance >= beforeBalance,
+            "Reward contract has lost tokens"
+        );
+
+        uint256 addedBalance = currentBalance.sub(beforeBalance);
+
+        totalRewards = totalRewards.add(addedBalance);
+        unallocatedRewards = unallocatedRewards.add(addedBalance);
+    }
+
+    /// @notice Sends the reward for a keep to the keep members.
+    /// @param keepIdentifier A unique identifier for the keep,
+    /// e.g. address or number converted to a `bytes32`.
+    function receiveReward(bytes32 keepIdentifier)
```

Comment:
Do we want to emit an event when revard is claimed? It should be cheap but can help us with some stats/debug later.